### PR TITLE
Blog databricks

### PR DESCRIPTION
**Pull request by mand35**
This Blog entry describes how to provision SDL on Databricks with a two stage download, select ingest pipeline into the Databricks database.
Another try

### CODE REVIEW

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+slug: sdl-databricks
+authors:
+  - name: Mandes Schönherr
+    title: Dr.sc.nat.
+    url: https://github.com/mand35
+tags: [Databricks, Cloud]
+hide_table_of_contents: false
+---
+
+Many analytics applications are ported to the cloud, Data Lakes and Datahouses in the cloud becoming more and more popular. The [Databricks](https://databricks.com) platform provides an easy accessible and easy configurable. SmartDataLake Builder on the other hand provides an open source, portable framework to load and transform the data.
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
Many analytics applications are ported to the cloud, Data Lakes and Lakehouses in the cloud becoming more and more popular. The [Databricks](https://databricks.com) platform provides an easy accessible and easy configurable. SmartDataLake Builder on the other hand provides an open source, portable automation tool to load and transform the data.
```

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+In this article the deployment of SmartDataLakeBuilder (SDLB) on [Databricks](https://databricks.com) is described. 
+
+<!--truncate-->
+
+Before jumping in, it should be mentioned, that there are also other methods to deploy SDLB in the cloud, e.g. using containers on Azure.
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
Before jumping in, it should be mentioned, that there are also many other methods to deploy SDLB in the cloud, e.g. using containers on Azure, Azure Kubernetes Service, Azure Synapse Clusters, ...
```

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    - Account and Workspace creation are described in detail [here](https://docs.databricks.com/getting-started/account-setup.html), there are few hints and modifications presented below.
+    - I selected AWS backend, but there are conceptual no differences to the other providers. If you already have an Azure, AWS or Google Cloud account/subscription this can be used, otherwise you can register a trial subscription there. 
+1. **Workspace stack** is created using the Quickstart as described in the documentation. When finished launch the Workspace.
+1. **Databricks CLI**: for file transfer of configuration files, scripts and data, the [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html) is installed locally. **Configure** the CLI, using the Workspace URL and in the Workspace "Settings" -> "User Settings" -> "Access tokens" create a new token.
+1. **Cluster** creation, in the Workspace open the *Cluster* Creation form.
+	- Spark version: When selecting the *Databricks version* pay attention to the related Spark version. 
+	  This need to match with the version we build SDLB later. Here `10.4 TLS` is selected with `Spark 3.2.1` and `Scala 2.12`.
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
	  This need to match with the Spark version we build SDLB later. Here `10.4 TLS` is selected with `Spark 3.2.1` and `Scala 2.12`. 
```

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    - I selected AWS backend, but there are conceptual no differences to the other providers. If you already have an Azure, AWS or Google Cloud account/subscription this can be used, otherwise you can register a trial subscription there. 
+1. **Workspace stack** is created using the Quickstart as described in the documentation. When finished launch the Workspace.
+1. **Databricks CLI**: for file transfer of configuration files, scripts and data, the [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html) is installed locally. **Configure** the CLI, using the Workspace URL and in the Workspace "Settings" -> "User Settings" -> "Access tokens" create a new token.
+1. **Cluster** creation, in the Workspace open the *Cluster* Creation form.
+	- Spark version: When selecting the *Databricks version* pay attention to the related Spark version. 
+	  This need to match with the version we build SDLB later. Here `10.4 TLS` is selected with `Spark 3.2.1` and `Scala 2.12`. 
+	  Alternatively, SDL could be build with a different version.
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
	  Alternatively, SDLB can be build with a different Spark versions, see also [Architecture](../../docs/architecture) for supported versions. 
```

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+1. **Workspace stack** is created using the Quickstart as described in the documentation. When finished launch the Workspace.
+1. **Databricks CLI**: for file transfer of configuration files, scripts and data, the [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html) is installed locally. **Configure** the CLI, using the Workspace URL and in the Workspace "Settings" -> "User Settings" -> "Access tokens" create a new token.
+1. **Cluster** creation, in the Workspace open the *Cluster* Creation form.
+	- Spark version: When selecting the *Databricks version* pay attention to the related Spark version. 
+	  This need to match with the version we build SDLB later. Here `10.4 TLS` is selected with `Spark 3.2.1` and `Scala 2.12`. 
+	  Alternatively, SDL could be build with a different version. 
+	- typesafe version correction script: the workspace currently brings typesafe:config 1.2.1 with it. 
+	  SDLB relies on functions of newer typesafe version (>1.3.0). 
+	  Thus, we provide a newer typesafe version in an initialization script: *Advanced options* -> *Init Scripts* specify `dbfs:/databricks/scripts/config-install.sh`
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
	- typesafe library version correction script: the workspace currently includes version 1.2.1 from com.typesafe:config java library. 
	  SDLB relies on functions of a newer version (>1.3.0) of this library. 
	  Thus, we provide a newer version of the com.typesafe:config java library in an initialization script: *Advanced options* -> *Init Scripts* specify `dbfs:/databricks/scripts/config-install.sh`
```

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+		+ Further, the script need to be created and uploaded. I used the following commands in local Terminal:
+		```
+		cat << EOF >> ./config-install.sh
+		#!/bin/bash
+		wget -O /databricks/jars/-----config-1.4.1.jar https://repo1.maven.org/maven2/com/typesafe/config/1.4.1/config-1.4.1.jar
+		EOF
+		databricks fs mkdirs dbfs:/databricks/scripts
+		databricks fs cp ./config-install.sh dbfs:/databricks/scripts/
+		```
+		Note: to double check the typesafe version I ran `grep typesafe pom.xml` in the [SmartDataLake](https://github.com/smart-data-lake/smart-data-lake.git) source
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
		Note: to double check the library version I ran `grep typesafe pom.xml` in the [SmartDataLake](https://github.com/smart-data-lake/smart-data-lake.git) source
```

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+		Note: the added `-----` will verify that this `.jar` is preferred before the default Workspace Spark version (which starts with `----`). 
+		If you are curious you could double check e.g. with a Workspace Python Notebook running `!ls /databricks/jars/*config*`
+
+1. **fat-jar**:
+       We need to provide the SDL sources and all required libraries. Therefore, we compile and pack the Scala code into a Jar including the dependencies. We use the [getting-started](https://github.com/smart-data-lake/getting-started.git) as dummy project, which itself pulls the SDL sources.
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
       We need to provide the SDLB sources and all required libraries. Therefore, we compile and pack the Scala code into a Jar including the dependencies. We use the [getting-started](https://github.com/smart-data-lake/getting-started.git) as dummy project, which itself pulls the SDLB sources. 
```

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+		Note: the added `-----` will verify that this `.jar` is preferred before the default Workspace Spark version (which starts with `----`). 
+		If you are curious you could double check e.g. with a Workspace Python Notebook running `!ls /databricks/jars/*config*`
+
+1. **fat-jar**:
+       We need to provide the SDL sources and all required libraries. Therefore, we compile and pack the Scala code into a Jar including the dependencies. We use the [getting-started](https://github.com/smart-data-lake/getting-started.git) as dummy project, which itself pulls the SDL sources. 
+	- download the [getting-started](https://github.com/smart-data-lake/getting-started.git) source and build it with the `-P fat-jar` profile
+	```
+    podman run -v ${PWD}:/mnt/project -v ${PWD}/.mvnrepo:/mnt/.mvnrepo maven:3.6.0-jdk-11-slim -- mvn -DskipTests  -P fat-jar  -f /mnt/project/pom.xml "-Dmaven.repo.local=/mnt/.mvnrepo" package
+	```
+	General build instructions can be fond in the [getting-started](../../docs/getting-started/setup#compile-scala-classes) documentation.
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
	General build instructions can be found in the [getting-started](../../docs/getting-started/setup#compile-scala-classes) documentation. 
```

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+	```
+    podman run -v ${PWD}:/mnt/project -v ${PWD}/.mvnrepo:/mnt/.mvnrepo maven:3.6.0-jdk-11-slim -- mvn -DskipTests  -P fat-jar  -f /mnt/project/pom.xml "-Dmaven.repo.local=/mnt/.mvnrepo" package
+	```
+	General build instructions can be fond in the [getting-started](../../docs/getting-started/setup#compile-scala-classes) documentation. 
+	Therewith, the file `target/getting-started-1.0-jar-with-dependencies.jar` is created. 
+	The *fat-jar* profile will include all required dependencies. The profile is defined in the [smart-data-lake](https://github.com/smart-data-lake/smart-data-lake) pom.xml
+
+1. upload files
+	- JAR: in the "Workspace" -> your user -> create a directory `jars` and "import" the library using the link in "(To import a library, such as a jar or egg, click here)" and select the above created fat-jar to upload. As a result the jar will be listed in the Workspace directory. 
+	- **SDLB application**: As an example a dataset from Airbnb NYC will be downloaded from Github, first written into a CSV file and later partially ported into a database. Therefore, the pipeline is defined first locally in a new file `application.conf`:
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
	- **SDLB application**: As an example a dataset from Airbnb NYC will be downloaded from Github, first written into a CSV file and later partially ported into a table. Therefore, the pipeline is defined first locally in a new file `application.conf`:
```

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+	Add a second job task (after creation, click the plus below) for the "copy" feed using the same specifications, but parameters `["-c", "file:///dbfs/conf/", "--feed-sel", "copy"]`
+	![download task](copy_task.png)
+
+1. **Launch** the job: 
+	When finished in the "Runs" in that job you should see the successful run status
+	![job status](job_status.png) 
+
+1. **Results**
+	After running the SDLB pipeline the data should be downloaded into the staging file and selected parts into the database:
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
	After running the SDLB pipeline the data should be downloaded into the staging file and selected parts into the table `stg_ab:`
```

File: blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+	- csv file: in the first step we downloaded the CSV file. This can be verified, e.g. by inspecting the data directory in the Databricks CLI using `databricks fs ls dbfs:/data/stg-ab` or running in a Workspace Python notebook `!ls /dbfs/data/stg-ab`
+	- database: in the second phase specific columns are put into the database. This can be verified in the Workspace -> Data -> default -> stg_ab
+	![select table](select_table.png)
+	![table](table.png)
+
+## Lessons Learned
+There are a few steps necessary, including building and uploading SDB. Further, we need to be careful with the used versions of the underlying libraries. With these few steps we can reveal the power of SDLB and Databricks, creating a portable and reproducible pipeline into a Databricks Lakehouse.
```
Comment (zzeekk) [inline blog/2022-04-07-SDL_databricks/2022-04-07-Databricks.md]:
```suggestion
There are a few steps necessary, including building and uploading SDLB. Further, we need to be careful with the used versions of the underlying libraries. With these few steps we can reveal the power of SDLB and Databricks, creating a portable and reproducible pipeline into a Databricks Lakehouse.  
```

**Review by zzeekk — COMMENTED** [review]
Nice Post and good level of detail for a post.

**Comment by mand35** [comment]
Many thanks @zzeekk for the corrections.

**Review by zzeekk — APPROVED** [review]
Great, thanks!

**Review by pgruetter — CHANGES_REQUESTED** [review]
I've created a pull request with some proposals. 

General:
- As discussed, I think we should remove the need for 2 jobs in this example as it gives a wrong impression
- Probably from the examples, but why is the table called `stg_ab` for a DataObject `int_ab` ?

**Comment by mand35** [comment]
Thanks both. I merged @pgruetter suggestions, removed the second job step and cleaned it.

**Pull request merged**